### PR TITLE
drivers: rtc: nucleo_wb55rg: Fix RTC callback issue when using power management config

### LINKED ISSUE #68673: drivers: rtc: nucleo_wb55rg: unable to setup RTC alarm callback after 1st exit from low-power mode
**Describe the bug**

I tried to implement Internal RTC using EVK [ST Nucleo WB55RG](https://docs.zephyrproject.org/latest/boards/arm/nucleo_wb55rg/doc/nucleo_wb55rg.html), but always fail when set the RTC time with return error code -EIO.

This is my main code:

```
#include <zephyr/drivers/rtc.h>
#include <zephyr/sys/timeutil.h>
#include <zephyr/device.h>

static const struct device *rtcDev = DEVICE_DT_GET(DT_ALIAS(rtc));

int main(void)
{
    struct rtc_time datetime_set;
    time_t timer_set = (1707208172UL);
    gmtime_r(&timer_set, (struct tm *)(&datetime_set));
    printk("Time: %02d:%02d:%02d\n", datetime_set.tm_hour, datetime_set.tm_min, datetime_set.tm_sec);
    printk("Date: %02d/%02d/%04d\n", datetime_set.tm_mday, datetime_set.tm_mon, datetime_set.tm_year);
    int err = rtc_set_time(rtcDev, &datetime_set);
    if(err == 0) {
        printk("Success to set time\n");
    } else {
        printk("Fail to set time: %i\n", err);
    }

    struct rtc_time datetime_get;
    err = rtc_get_time(rtcDev, &datetime_get);
    if(err == 0) {
        printk("Success to get time\n");
        printk("%02d:%02d:%02d\n", datetime_get.tm_hour, datetime_get.tm_min, datetime_get.tm_sec);
        printk("%02d/%02d/%04d\n", datetime_get.tm_mday, datetime_get.tm_mon, datetime_get.tm_year);
    } else {
        printk("Fail to get time: %i\n", err);
    }

    while (1) {
        k_sleep(K_SECONDS(5));
    }

    return 0;
}
```

I create my own board with name `stm32wb_sensor`. Here is the dts file:

```
/ {
	model = "STM32WB sensor board";
	compatible = "st,stm32wb55rg-nucleo";

	chosen {
		zephyr,console = &usart1;
		zephyr,shell-uart = &usart1;
		zephyr,bt-mon-uart = &usart1;
		zephyr,bt-c2h-uart = &usart1;
		zephyr,sram = &sram0;
		zephyr,flash = &flash0;
		zephyr,code-partition = &slot0_partition;
	};

	aliases {
		rtc = &rtc;
	};
};

&rtc {
	clocks = <&rcc STM32_CLOCK_BUS_APB1 0x00000400>,
	         <&rcc STM32_SRC_LSI RTC_SEL(2)>;
	status = "okay";

	backup_regs {
		status = "okay";
	};
};
```

Here is the final RTC node on `zephyr.dts` file on build folder:

```
		rtc: rtc@40002800 {
			compatible = "st,stm32-rtc";
			reg = < 0x40002800 0x400 >;
			interrupts = < 0x29 0x0 >;
			clocks = < &rcc 0x58 0x400 >, < &rcc 0x4 0x26890 >;
			prescaler = < 0x8000 >;
			status = "okay";
			bbram: backup_regs {
				compatible = "st,stm32-bbram";
				st,backup-regs = < 0x14 >;
				status = "okay";
			};
		};
``` 

Here is my `prj.conf` that related to RTC:

```
# RTC
CONFIG_RTC=y
```

**My Observation**

After digging down into the kernel driver, found out the -EIO is come from `zephyr/drivers/rtc/rtc_ll_stm32.c` on `rtc_stm32_enter_initialization_mode` function at line 87. I think `LL_RTC_IsActiveFlag_INIT(RTC)` never return true.

Aside from that, I found that `rtc_stm32_init` on the same file is never being called, even though it's registered using macro `DEVICE_DT_INST_DEFINE(0, &rtc_stm32_init, NULL, &rtc_data, &rtc_config, PRE_KERNEL_1, CONFIG_RTC_INIT_PRIORITY, &rtc_stm32_driver_api);` I suppose `rtc_stm32_init` should be called at least once.

**To Reproduce**

1. Create `docker_shell.sh`
```
DOCKER_ARGS="--privileged \
--rm=true \
--user="$(id --user):$(id --group)" \
-e ZEPHYR_SDK_INSTALL_DIR=/opt/toolchains/zephyr-sdk-0.16.5 \
-e CCACHE_DIR=$(pwd)/.ccache \
-v $(pwd):$(pwd) \
-w $(pwd)"
DOCKER_IMAGE=zephyrprojectrtos/ci:v0.26.7

docker run $DOCKER_ARGS $DOCKER_IMAGE "$@"
```
2. `./docker_shell.sh west build -b stm32wb_sensor --sysbuild <path_to_app>`
3. See error (see below on logs)

**Expected behavior**
When calling `rtc_set_time`, I expect the return value to be 0, not -5 (EIO).

**Impact**
It's blocking my task.

**Logs and console output**
```
Time: 08:29:32
Date: 06/01/0124
Fail to set time: -5
Success to get time
00:00:01
01/00/0100
```

**Environment (please complete the following information):**
 - OS: Host: Linux Ubuntu 18.04, Docker: zephyrprojectrtos/ci:v0.26.7
 - Toolchain: Zephyr SDK v0.16.5
 - Zephyr RTOS v3.5.0
 
I hope this information is complete enough to spark further idea. Any help or pointer to resolve the problem is pretty much appreciated. Thank you for your time and suggestion.
 
Krisna

### PR DESCRIPTION
**Pull request by krisnaresi**
It is found that when we use CONFIG_COUNTER and CONFIG_PM concurrently, the RTC alarm callback can be used only once (in some cases, it just won't work at all, e.g., using CONFIG_BT). By set the `DBP` bit on `PWR control register 1` via `LL_PWR_EnableBkUpAccess` function to temporarily disable write protection every time we assign RTC alarm, we can register alarm callback correctly. Tested on Nucleo WB55RG.

Fixes: #68673

### CODE REVIEW

**Review by erwango — CHANGES_REQUESTED** [review]

**Comment by ajarmouni-st** [comment]
@krisnaresi You have compliance issues in CI.

```
Commit 94f65adbd8:
1: UC5 Commit title exceeds max length (95>75): "drivers: rtc: nucleo_wb55rg: Fix RTC alarm when using CONFIG_COUNTER and CONFIG_PM concurrently"
3: UC4 Commit message body line exceeds max length (404>75): "It is found that when we use CONFIG_COUNTER and CONFIG_PM concurrently, the RTC alarm callback can be used only once (in some cases, it just won't work at all, e.g., using CONFIG_BT). By set the DBP bit on PWR control register 1 via LL_PWR_EnableBkUpAccess function to temporarily disable write protection every time we assign RTC alarm, we can register alarm callback correctly. Tested on Nucleo WB55RG."

Commit effdc[29](https://github.com/zephyrproject-rtos/zephyr/actions/runs/8246002409/job/22551936595?pr=70078#step:9:30)b77:
1: UC2 Commit message does not contain a 'Signed-off-by:' line
1: UC3 Commit title does not follow [subsystem]: [subject] (and should not start with literal subsys:): "Fix RTC alarm when using CONFIG_COUNTER and CONFIG_PM concurrently"
1: UC6 Commit message body is empty, should at least have 1 line(s).
3: B6 Body message is missing
Error: See https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-guidelines for more details

 94f65adbd8685b3ae56f7d9f0cecd7d6fdefb16b: author email (krisnaresi <krisna_resi@ymail.com>) needs to match one of the signed-off-by entries.
94f65adbd8685b3ae56f7d9f0cecd7d6fdefb16b: author email (krisnaresi <krisna_resi@ymail.com>) does not follow the syntax: First Last <email>.
 effdc29b77eedeb61679e5ce3de8b13d8b45[38](https://github.com/zephyrproject-rtos/zephyr/actions/runs/8246002409/job/22551936595?pr=70078#step:9:39)c3: author email (krisnaresi <krisna_resi@ymail.com>) needs to match one of the signed-off-by entries.
effdc29b77eedeb61679e5ce3de8b13d8b4538c3: author email (krisnaresi <krisna_resi@ymail.com>) does not follow the syntax: First Last <email>.
```

**Comment by krisnaresi** [comment]
> @krisnaresi You have compliance issues in CI.
> 
> ```
> Commit 94f65adbd8:
> 1: UC5 Commit title exceeds max length (95>75): "drivers: rtc: nucleo_wb55rg: Fix RTC alarm when using CONFIG_COUNTER and CONFIG_PM concurrently"
> 3: UC4 Commit message body line exceeds max length (404>75): "It is found that when we use CONFIG_COUNTER and CONFIG_PM concurrently, the RTC alarm callback can be used only once (in some cases, it just won't work at all, e.g., using CONFIG_BT). By set the DBP bit on PWR control register 1 via LL_PWR_EnableBkUpAccess function to temporarily disable write protection every time we assign RTC alarm, we can register alarm callback correctly. Tested on Nucleo WB55RG."
> 
> Commit effdc[29](https://github.com/zephyrproject-rtos/zephyr/actions/runs/8246002409/job/22551936595?pr=70078#step:9:30)b77:
> 1: UC2 Commit message does not contain a 'Signed-off-by:' line
> 1: UC3 Commit title does not follow [subsystem]: [subject] (and should not start with literal subsys:): "Fix RTC alarm when using CONFIG_COUNTER and CONFIG_PM concurrently"
> 1: UC6 Commit message body is empty, should at least have 1 line(s).
> 3: B6 Body message is missing
> Error: See https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-guidelines for more details
> 
>  94f65adbd8685b3ae56f7d9f0cecd7d6fdefb16b: author email (krisnaresi <krisna_resi@ymail.com>) needs to match one of the signed-off-by entries.
> 94f65adbd8685b3ae56f7d9f0cecd7d6fdefb16b: author email (krisnaresi <krisna_resi@ymail.com>) does not follow the syntax: First Last <email>.
>  effdc29b77eedeb61679e5ce3de8b13d8b45[38](https://github.com/zephyrproject-rtos/zephyr/actions/runs/8246002409/job/22551936595?pr=70078#step:9:39)c3: author email (krisnaresi <krisna_resi@ymail.com>) needs to match one of the signed-off-by entries.
> effdc29b77eedeb61679e5ce3de8b13d8b4538c3: author email (krisnaresi <krisna_resi@ymail.com>) does not follow the syntax: First Last <email>.
> ```

@ajarmouni-st, I think I have resolve commit message. Still waiting for re-run on the workflow. Let me know if you spot something. Thanks.

**Comment by ajarmouni-st** [comment]
The good news, I run `tests/drivers/counter/counter_basic_api/drivers.counter.basic_api` with `CONFIG_PM=y` & all the tests pass with this patch. Previously the tests `test_single_shot_alarm_notop` & `test_all_channels` were failing for RTC counter.

So it would be great if you can add a commit with the following test case in `tests/drivers/counter/counter_basic_api/testcase.yaml`
```
  drivers.counter.basic_api.stm32_pm:
    tags:
      - drivers
      - counter
      - power
    depends_on: counter
    min_ram: 16
    platform_allow:
      - nucleo_wb55rg
    extra_configs:
      - CONFIG_PM=y
    timeout: 600
```

**Comment by ajarmouni-st** [comment]
Update: `test_single_shot_alarm_notop` is still failing with `CONFIG_PM=y` ...

**Comment by krisnaresi** [comment]
Hi @ajarmouni-st, this morning I already retest on my project with the exact main branch commit [b47c66e](https://github.com/krisnaresi/zephyr/commit/b47c66e3f293c171e388fb285ae4f30e31068e62) and make changes as per this pull request. Tested using source code that I shared with [you](https://github.com/zephyrproject-rtos/zephyr/issues/68673#issuecomment-1945864671) and with those [prj.conf](https://github.com/zephyrproject-rtos/zephyr/issues/68673#issuecomment-1972473821). It's work properly on my side.

Now, I don't know how to do ZTest, maybe you can share the step and build command on how to reproduce `test_single_shot_alarm_notop` error to me? And I see there is [overlay](https://github.com/krisnaresi/zephyr/blob/main/tests/drivers/counter/counter_basic_api/boards/nucleo_wb55rg.overlay) here with a lot of timers, do you think it can cause this problem?

**Comment by ajarmouni-st** [comment]
> Now, I don't know how to do ZTest, maybe you can share the step and build command on how to reproduce `test_single_shot_alarm_notop` error to me? And I see there is [overlay](https://github.com/krisnaresi/zephyr/blob/main/tests/drivers/counter/counter_basic_api/boards/nucleo_wb55rg.overlay) here with a lot of timers, do you think it can cause this problem?

You can run the test with
`west build -p always -b nucleo_wb55rg -T tests/drivers/counter/counter_basic_api/drivers.counter.basic_api -- -DCONFIG_PM=y`

In any case, since the RTC counter tests are not all passing with `CONFIG_PM=y`, adding a test case for it isn't possible right now & can be left for another PR when the problem is fixed.

**Comment by krisnaresi** [comment]
Hi @ajarmouni-st, yes I confirm that `test_single_shot_alarm_notop` is not working on my side too. Seems like my fix can only get `test_all_channel` to work with `CONFIG_PM=y`. I already tried several things, none of them can make `test_single_shot_alarm_notop` to works.

I understand that my PR is not fixing everything, so I don't mind if I should retract my PR. Please let me know the final decision for this PR, so I can close both PR and my issue. In any case, our team will use `LL_PWR_EnableBkUpAccess()` for time being until generic solution is available by ST teams. Thanks.

**Comment by erwango** [comment]
> Hi @ajarmouni-st, yes I confirm that `test_single_shot_alarm_notop` is not working on my side too. Seems like my fix can only get `test_all_channel` to work with `CONFIG_PM=y`. I already tried several things, none of them can make `test_single_shot_alarm_notop` to works.
> 
> I understand that my PR is not fixing everything, so I don't mind if I should retract my PR. Please let me know the final decision for this PR, so I can close both PR and my issue. In any case, our team will use `LL_PWR_EnableBkUpAccess()` for time being until generic solution is available by ST teams. Thanks.

One last idea to test would be to align implementation on driver/rtc/rtc_stm32.c which seem more complete from BkUp Access point of view. If this doesn't help solving the last isue, I don't mind merging as is, as this fixes some bugs at least.

**Review by ajarmouni-st — CHANGES_REQUESTED** [review]

File: drivers/counter/counter_ll_stm32_rtc.c
Comment (ajarmouni-st) [reply drivers/counter/counter_ll_stm32_rtc.c:247]: @krisnaresi I'am ready to approve the PR, just need this change to be taken into account. Thanks for your collaboration!

**Comment by krisnaresi** [comment]
> > Hi @ajarmouni-st, yes I confirm that `test_single_shot_alarm_notop` is not working on my side too. Seems like my fix can only get `test_all_channel` to work with `CONFIG_PM=y`. I already tried several things, none of them can make `test_single_shot_alarm_notop` to works.
> > I understand that my PR is not fixing everything, so I don't mind if I should retract my PR. Please let me know the final decision for this PR, so I can close both PR and my issue. In any case, our team will use `LL_PWR_EnableBkUpAccess()` for time being until generic solution is available by ST teams. Thanks.
> 
> One last idea to test would be to align implementation on driver/rtc/rtc_stm32.c which seem more complete from BkUp Access point of view. If this doesn't help solving the last isue, I don't mind merging as is, as this fixes some bugs at least.

Hi @erwango, I already implement `LL_PWR_EnableBkUpAccess()` on `rtc_stm32_get_time` function on `drivers/rtc/rtc_ll_stm32.c`, but `test_single_shot_alarm_notop` test still failed. As I recall, `rtc_ll_stm32.c` still not implement alarm yet, so in my opinion, I don't think it can solve any test alarm problem.

File: drivers/counter/counter_ll_stm32_rtc.c
Comment (krisnaresi) [reply drivers/counter/counter_ll_stm32_rtc.c:247]: @ajarmouni-st, I think you're correct, let me push new commit for that. Thanks.

**Comment by ajarmouni-st** [comment]
@krisnaresi no I just run the tests w/t this change, they pass. No need to add anything.

**Review by ajarmouni-st — APPROVED** [review]

**Comment by krisnaresi** [comment]
> @krisnaresi no I just run the tests w/t this change, they pass. No need to add anything.

Oh, ok then. So, it's ready to merge I suppose? I'm not sure if I should be the one to merge it or somebody else that have write access..

**Comment by ajarmouni-st** [comment]
> Oh, ok then. So, it's ready to merge I suppose? 

Pending a second approval

> somebody else that have write access.

Yes

**Review by erwango — APPROVED** [review]

**Pull request merged**
